keps/sig-cloud-provider/1959-service-lb-class-annotation/README.md

```diff
@@ -0,0 +1,418 @@
+
+Introduce a new Service annotation `service.kubernetes.io/load-balancer-name`.
+
+If the loadbalancer class annotation is not set, the existing cloud provider
+will assume ownership of the Service Type=LoadBalancer resource. This is required
+to not break existing clusters that assume Service Type=LoadBalancer is always
+managed by the cloud provider.
+
+A new method `SupportedLoadBalancerClassNames()` will be added to the cloud provider
+`LoadBalancer` interface which returns a list of strings representing the
+class names that the cloud provider supports.
+
+```go
+type LoadBalancer interface {
+	GetLoadBalancer(ctx context.Context, clusterName string, service *v1.Service) (status *v1.LoadBalancerStatus, exists bool, err error)
+	GetLoadBalancerName(ctx context.Context, clusterName string, service *v1.Service) string
+	EnsureLoadBalancer(ctx context.Context, clusterName string, service *v1.Service, nodes []*v1.Node) (*v1.LoadBalancerStatus, error)
+	UpdateLoadBalancer(ctx context.Context, clusterName string, service *v1.Service, nodes []*v1.Node) error
+	EnsureLoadBalancerDeleted(ctx context.Context, clusterName string, service *v1.Service) error
+	SupportedLoadBalancerClassNames() []string
+}
+```
+
+Required updates to service controller:
+* if the class annotation is NOT set for a Service, allow the cloud provider
+to reconcile the load balancer.
+* if the class annotation IS set for a Service, call `SupportedLoadBalancerClassNames`
+from the cloud provider and check if the class name matches any of the supported class names.
+* if the class annotation IS set for a Service, but `SupportedLoadbalancerClassNames`
+does not return any class names that match the annotation, then skip reconciliation of the Service.
+
+### Test Plan
+
+Unit tests:
+* test that service controller does not call the cloud provider if the class annotation is set and the cloud provider does not support it.
+* test that service controller calls the cloud provider if one of the supported classes from `SupportedLoadBalancerClassNames` is set.
+* the annotation `service.kubernetes.io/load-balancer-class` is not accepted when the feature gate `ServiceLoadBalancerClass` is disabled.
+
+E2E tests:
+* test that creating a Service with an unknown class annotation results in no load balancer being created for a Service.
+
+### Graduation Criteria
+
+N/A since we can't apply alpha/beta/GA criteria for annotations.
+
+### Upgrade / Downgrade Strategy
+
+On upgrade, use of this annotation will be allowed. On downgrade, service controller
+may ignore existing Services with the annotation, leading to multiple implementations
+trying to create load balancers. On downgrade, if the class annotation is used
+and there are multiple implementations of Service Type=LoadBalancer, a user must ensure
+there is only 1 implementation of Service Type=LoadBalancer in the cluster.
+
+Though the downgrade scenario isn't ideal, it is assumed if that a cluster was upgraded to v1.20,
```

Comment:
On the fence about this -- if you downgrade a cluster that is using the class annotation and there are multiple Service Type=LoadBalancer implementations, you would end up with the cloud provider fighting with other implementations on downgrade. Unlike fields, we can't drop the annotation to ensure safe rollback. We _could_ add a feature gate, but that doesn't really fix the downgrade scenario. 

Also, downgrading a cluster after enabling a bunch of Service Type=LB implementations seems unlikely?